### ISSUE TITLE
Deadlock in BuildEventServiceTransport.java

### ISSUE THREAD

**Issue by EricBurnett** [bazelbuild/bazel#5402]
### Description of the problem / feature request:

Bazel can deadlock while writing to the BES. The issue is that sendBuildEvent() is reentrant, which in turn can cause multiple lock orderings. One such ordering pair is PrintStream causing a BuildEvent to be sent, and a BuildEvent being sent causing a file digest to be computed, causing a PrintStream to be called if it is slow.

The first path:

	at com.google.devtools.build.lib.buildeventservice.BuildEventServiceTransport.sendBuildEvent(BuildEventServiceTransport.java:289)
	- waiting to lock <0x000000070139db90> (a com.google.devtools.build.lib.buildeventservice.BuildEventServiceTransport)
	at com.google.devtools.build.lib.runtime.BuildEventStreamer.flush(BuildEventStreamer.java:540)
	at com.google.devtools.build.lib.runtime.SynchronizedOutputStream.write(SynchronizedOutputStream.java:129)
	at com.google.devtools.build.lib.util.io.DelegatingOutErr$DelegatingOutputStream.write(DelegatingOutErr.java:103)
	at com.google.devtools.build.lib.util.AnsiStrippingOutputStream.write(AnsiStrippingOutputStream.java:67)
	- locked <0x000000070139fdd0> (a com.google.devtools.build.lib.util.AnsiStrippingOutputStream)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
	at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
	- locked <0x000000070139fdb8> (a java.io.BufferedOutputStream)
	at com.google.devtools.build.lib.runtime.ExperimentalEventHandler.handle(ExperimentalEventHandler.java:365)
	- locked <0x000000070139fd20> (a com.google.devtools.build.lib.runtime.ExperimentalEventHandler)
	at com.google.devtools.build.lib.events.Reporter.handle(Reporter.java:113)
	- locked <0x000000070139fca8> (a com.google.devtools.build.lib.events.Reporter)
	at com.google.devtools.build.lib.events.ReporterStream.write(ReporterStream.java:55)
	at java.io.PrintStream.write(PrintStream.java:480)
	- locked <0x00000007019ca958> (a java.io.PrintStream)
	at sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:221)
	at sun.nio.cs.StreamEncoder.implFlushBuffer(StreamEncoder.java:291)
	at sun.nio.cs.StreamEncoder.flushBuffer(StreamEncoder.java:104)
	- locked <0x00000007019ca9a0> (a java.io.OutputStreamWriter)
	at java.io.OutputStreamWriter.flushBuffer(OutputStreamWriter.java:185)
	at java.io.PrintStream.newLine(PrintStream.java:546)
	- locked <0x00000007019ca958> (a java.io.PrintStream)
	at java.io.PrintStream.println(PrintStream.java:807)
	- locked <0x00000007019ca958> (a java.io.PrintStream)
	at com.google.devtools.build.lib.actions.cache.DigestUtils.getDigestInternal(DigestUtils.java:162)

The second path:

	at java.io.PrintStream.println(PrintStream.java:805)
	- waiting to lock <0x00000007019ca958> (a java.io.PrintStream)
	at com.google.devtools.build.lib.actions.cache.DigestUtils.getDigestInternal(DigestUtils.java:162)
	at com.google.devtools.build.lib.actions.cache.DigestUtils.getDigestOrFail(DigestUtils.java:249)
	at com.google.devtools.build.lib.remote.util.DigestUtil.compute(DigestUtil.java:64)
	at com.google.devtools.build.lib.remote.RemoteModule$CasPathConverter.apply(RemoteModule.java:69)
	at com.google.devtools.build.lib.runtime.ServerBuilder$1.apply(ServerBuilder.java:101)
	at com.google.devtools.build.lib.runtime.NamedArtifactGroup.asStreamProto(NamedArtifactGroup.java:63)
	at com.google.devtools.build.lib.buildeventservice.BuildEventServiceTransport.sendBuildEvent(BuildEventServiceTransport.java:289)
	- locked <0x000000070139db90> (a com.google.devtools.build.lib.buildeventservice.BuildEventServiceTransport)
	at com.google.devtools.build.lib.runtime.BuildEventStreamer.post(BuildEventStreamer.java:273)
	at com.google.devtools.build.lib.runtime.BuildEventStreamer.maybeReportArtifactSet(BuildEventStreamer.java:407)
	at com.google.devtools.build.lib.runtime.BuildEventStreamer.maybeReportArtifactSet(BuildEventStreamer.java:405)
	at com.google.devtools.build.lib.runtime.BuildEventStreamer.maybeReportArtifactSet(BuildEventStreamer.java:411)
	at com.google.devtools.build.lib.runtime.BuildEventStreamer.buildEvent(BuildEventStreamer.java:479)

I'm unsure what the correct fix is. This specific scenario can be avoided by removing the log from getDigestInternal, but that may not be the only path in which different lock orderings can arise. Probably better would be for BuildEventServiceTransport to not call anything that might produce a log line while in synchronized code, and then also handle the case of new build events being produced while preparing one to be sent.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

At a guess: do a build with many large input files, high --jobs, on a machine with a slow disk and/or severely CPU constrained, to trigger that specific log message to be hit a lot. Then hope you get lucky.

### What operating system are you running Bazel on?

Linux.

### What's the output of `bazel info release`?

Pulled from java.log instead: 0.13.1-201805171359+f41dd9a

###  Have you found anything relevant by searching the web?

Nope.

### Any other information, logs, or outputs that you want to share?

I have copies of the java.log and jstack output if need be, but I'm not at liberty to publically share them. Googlers: poke me if you want to see them, or more context on the exact scenario in which this was found. The person who actually found this bug (I'm just reporting it) encounters it with enough regularity that more details from a hung bazel can probably be extracted if necessary.

**Comment by brian-peloton** [comment]
I found this originally. For reference, there's nothing that seems relevant in our bazel version vs 0.13.1, just some cherry picks related to remote execution a few unrelated custom patches:
```console
$ git log --pretty=short f19d2c1..f41dd9a
commit f41dd9a97ce7593fe6ccb33dd6e096470a23e793
Author: Brian Silverman <bsilver16384@gmail.com>

    Expose feature-based flags from CROSSTOOL to Skylark.

commit 67326e9f8d7411b9c9ae94b4684f785cb299c429
Author: Brian Silverman <bsilver16384@gmail.com>

    Make CppCompileAction's extra action info deterministic

commit 67f8751501ec8a2c45acb4ef0a44a734560872c0
Author: Austin Schuh <austin.linux@gmail.com>

    Don't allow network access by default

commit 6f7069db05eff155d7aa11f3d07a757331d1196e
Author: James Judd <james@lucidchart.com>

    Sort entries by segment when building a parent node to prevent unordered directory structures.

commit caa49f28d5290b9d1e8b592f5e4d800aafba10ab
Author: Laszlo Csomor <laszlocsomor@google.com>

    test-setup: remove leading "./" from test name

commit 5653bf0bd54354e0773d5d9bc335f777d1696172
Author: asteinb <asteinb@google.com>

    Add error message on empty public resources

commit 0156642fbf577a2bf93092f6a608538017d0e799
Author: buchgr <buchgr@google.com>

    Automated rollback of commit 4465dae23de989f1452e93d0a88ac2a289103dd9.

commit 8812669bd601fc9e8d71f94d2058eee2b6399187
Author: Laszlo Csomor <laszlocsomor@google.com>

    windows: GetOutputRoot() returns GetHomeDir()
```

**Comment by ulfjack** [comment]
The exact stack trace provided here can no longer happen at head as of ce9aafcf63dd198fe16c682dfeedc444f07b5669, but I can't say for sure that we don't have any remaining issues left.

**Comment by ulfjack** [comment]
A potential deadlock in the BEP/BES is a P1 issue, not P3. We're about to roll it out to all of Google, so this is a problem.

**Comment by ulfjack** [comment]
Btw. it's really, really expensive for us to re-digest the files in CasPathConverter - that might be a serious performance issue.

**Comment by AustinSchuh** [comment]
Any update?  We are now starting to hit this more reliably.

**Comment by ulfjack** [comment]
More stack traces of deadlocks would help.

**Comment by buchgr** [comment]
I am currently looking into this.

**Comment by buchgr** [comment]
I think we have enough information to know what's happening. 

T1 | T2 
--- | ---
sends a build event (lock1) | Writes stderr (lock2) ...
Writes a "Slow Read:" warning (lock2) | ... and realizes that the stderr buffer is full, thus flushes and sends a build event (lock1).

We can fix this particular case, but I need some more thinking on how to ensure that some other case doesn't trigger a deadlock too.

**Comment by buchgr** [comment]
So in 59f9159 I have just removed the lock on `BuildEventServiceTransport.sendBuildEvent`. That hasn't fix the bug, because the bug had already been fixed as a side effect of ce9aafc as we moved the serialization of the build event protos from the `sendBuildEvent` method to the uploader thread.
